### PR TITLE
Fix clienthellotest to use PACKET functions

### PR DESCRIPTION
**Pull request by dwmw2**
I copied the clienthellotest to make bad_dtls_test. You told me I should be using the `PACKET` functions. Let's fix clienthellotest to stop being a bad example. Or should it be client_hello_test... ? :)

### CODE REVIEW

File: ssl/packet_locl.h [hunk 1]
```diff
@@ -102,6 +102,23 @@ __owur static ossl_inline int PACKET_equal(const PACKET *pkt, const void *ptr,
 }
 
 /*
+ * Similar to PACKET_equal() except that it compares only a limited number
+ * of bytes (i.e. as strncmp() is to strcmp()). And that it consumes the
+ * compared bytes on success.
+ */
+__owur static ossl_inline int PACKET_nequal(PACKET *pkt, const void *ptr,
```
Comment (mattcaswell) [inline ssl/packet_locl.h]: Not sure about the name. Perhaps `PACKET_equaln` might be better - I automatically read "nequal" as "not equal".

File: test/clienthellotest.c [hunk 2]
```diff
@@ -81,50 +82,50 @@ int main(int argc, char *argv[])
...
-        data += CLIENT_VERSION_LEN + SSL3_RANDOM_SIZE;
-        if (data + SESSION_ID_LEN_LEN > dataend)
+        if (!PACKET_forward(&pkt, CLIENT_VERSION_LEN + SSL3_RANDOM_SIZE))
             goto end;
+
         /* Skip session id */
-        tmplen = *data;
-        data += SESSION_ID_LEN_LEN + tmplen;
-        if (data + CIPHERS_LEN_LEN > dataend)
+        if (!PACKET_get_1(&pkt, &tmplen) || !PACKET_forward(&pkt, tmplen))
```
Comment (mattcaswell) [inline test/clienthellotest.c]: `PACKET_get_length_prefixed_1` would probably be a better choice here.

File: test/clienthellotest.c [hunk 3]
```diff
@@ -81,50 +82,50 @@ int main(int argc, char *argv[])
...
             goto end;
+
         /* Skip ciphers */
-        tmplen = ((*data) << 8) | *(data + 1);
-        data += CIPHERS_LEN_LEN + tmplen;
-        if (data + COMPRESSION_LEN_LEN > dataend)
+        if (!PACKET_get_net_2(&pkt, &tmplen) || !PACKET_forward(&pkt, tmplen))
```
Comment (mattcaswell) [inline test/clienthellotest.c]: `PACKET_get_length_prefixed_2`

File: test/clienthellotest.c [hunk 4]
```diff
@@ -81,50 +82,50 @@ int main(int argc, char *argv[])
...
             goto end;
+
         /* Skip compression */
-        tmplen = *data;
-        data += COMPRESSION_LEN_LEN + tmplen;
-        if (data + EXTENSIONS_LEN_LEN > dataend)
+        if (!PACKET_get_1(&pkt, &tmplen) || !PACKET_forward(&pkt, tmplen))
```
Comment (mattcaswell) [inline test/clienthellotest.c]: `PACKET_get_length_prefixed_1`

File: test/clienthellotest.c [hunk 5]
```diff
@@ -81,50 +82,50 @@ int main(int argc, char *argv[])
...
             goto end;
+
         /* Extensions len */
-        tmplen = ((*data) << 8) | *(data + 1);
-        data += EXTENSIONS_LEN_LEN;
-        if (data + tmplen > dataend)
+        if (!PACKET_get_net_2(&pkt, &tmplen))
+            goto end;
+        if (tmplen != PACKET_remaining(&pkt))
```
Comment (mattcaswell) [inline test/clienthellotest.c]: `PACKET_as_length_prefixed_2`

File: test/clienthellotest.c [hunk 6]
```diff
@@ -81,50 +82,50 @@ int main(int argc, char *argv[])
...
-        while (tmplen > EXTENSION_TYPE_LEN + EXTENSION_SIZE_LEN) {
-            type = ((*data) << 8) | *(data + 1);
-            data += EXTENSION_TYPE_LEN;
-            size = ((*data) << 8) | *(data + 1);
-            data += EXTENSION_SIZE_LEN;
-            if (data + size > dataend)
+        while (PACKET_remaining(&pkt)) {
+
+            if (!PACKET_get_net_2(&pkt, &type) ||
+                !PACKET_get_net_2(&pkt, &size))
```
Comment (mattcaswell) [inline test/clienthellotest.c]: `PACKET_get_length_prefixed_2`

**Comment by dwmw2** [comment]
Actually, I could rewrite all my uses of `PACKET_equaln()` to use `PACKET_get_length_prefixed_x() && PACKET_equal()`. So we don't need the new function at all.

File: test/clienthellotest.c [hunk 7]
```diff
@@ -81,50 +74,47 @@ int main(int argc, char *argv[])
...
-        while (tmplen > EXTENSION_TYPE_LEN + EXTENSION_SIZE_LEN) {
-            type = ((*data) << 8) | *(data + 1);
-            data += EXTENSION_TYPE_LEN;
-            size = ((*data) << 8) | *(data + 1);
-            data += EXTENSION_SIZE_LEN;
-            if (data + size > dataend)
+        while (PACKET_remaining(&pkt2)) {
+
+            if (!PACKET_get_net_2(&pkt2, &type) ||
+                !PACKET_get_length_prefixed_2(&pkt2, &pkt))
```
Comment (mattcaswell) [inline test/clienthellotest.c]: Perhaps using `pkt3` might be clearer here rather than reusing `pkt`?

Comment (dwmw2) [inline test/clienthellotest.c]: It might. Do you realise *just* how far down the road I've come from my initial _"oh, DTLS1_BAD_VER broke again..."_ ? :)

**Comment by mattcaswell** [comment]
+1

**Comment by mattcaswell** [comment]
Ping @richsalz - needs your review again. This has changed significantly since you originally reviewed it.

**Comment by richsalz** [comment]
way cool much better and cleaner.
+1

**Comment by mattcaswell** [comment]
Merged. Thanks!

**Pull request closed without merging**